modules/apigee_edge_apidocs/src/Entity/ApiDoc.php

```diff
@@ -231,7 +278,35 @@ public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
       ->setLabel(t('Changed'))
       ->setDescription(t('The time that the entity was last edited.'));
 
+    $fields['fetched_timestamp'] = BaseFieldDefinition::create('timestamp')
+      ->setLabel(t('Spec fetched from URL timestamp'))
+      ->setDescription(t('When the OpenAPI spec file was last fetched from URL as a Unix timestamp.'));
+
     return $fields;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function preSave(EntityStorageInterface $storage) {
+    parent::preSave($storage);
+
+    \Drupal::service('apigee_edge_apidocs.spec_fetcher')->fetchSpec($this, TRUE);
```

Comment:
Still, do this in the entity stroage and not in the entity object, you can use DI there. Override `\Drupal\Core\Entity\EntityStorageBase::doPreSave()` or \Drupal\Core\Entity\EntityStorageBase::save()`

Reply:
I don't think it's creating a new storage handler just for that.

Reply:
What is the problem with that? The storage handlers suppose to handle entity CRUD operations this is the reason why DI is available there and not in the entity classes that should be value objects only.